### ISSUE TITLE
Need help with skip frame

### ISSUE THREAD

**Issue by zhenghongzhi** [Unity-Technologies/ml-agents#293]
I'm training a game like this
![](https://raw.githubusercontent.com/zhenghongzhi/Unity-JumpJump/master/demo.gif)
The duration of player in the air varies from 15 frames to 200 frames and after falling on the ground should the player accepts actions.
**Is there a better way than setting Frame to Skip to 200 which wastes lots of frames?** 

I try to understand the code and realize it is now frame based to Step and DecideAction in FixedUpdate. It seems that adding manual Step and DecideAction is the solution?

Edit:
I find in Academy.cs line:345 that following code is not surrounded by `if (skippingFrames == false)`.  
Am I misunderstanding something?
```
AcademyStep();

foreach (Brain brain in brains)
{
     brain.Step();
}
```

**Comment by awjuliani** [comment]
Hi @zhenghongzhi,

Your project is very interesting! We will soon be adding a feature to selectively make decisions and actions for your agents. For now, it is possible to unsubscribe an agent from the brain between decision points, and then re-subscribe when it is time to act. 

Are you learning from images? Or from internal representation?

**Comment by zhenghongzhi** [comment]
Hi @awjuliani ,

Thank you for your help. I'm learning from internal representation. I will try to unsubscribe and re-subscribe for now and looking forward to the selective actions and decisions feature. 

And thank you for the awesome project.

**Comment by zhenghongzhi** [comment]
Hi @awjuliani ,
Unsubscribe and re-subscribe brain is now working with multiple game instance running at the same time.

**Comment by zhenghongzhi** [comment]
Hi @awjuliani  ,
I find it is impossible to unsubscribe and re-subscribe brain for now with multiple game instance running at the same time.
```
self.add_experiences(info, new_info, epsi, actions, a_dist, value)
```
agents in `info` and `new_info` may be not the same length.

**Comment by kwea123** [comment]
Why not just skip those infos where the player is not on the ground?
Like
```
if player_on_the_ground:
    self.add_experiences(info, new_info, epsi, actions, a_dist, value)
````

**Comment by zhenghongzhi** [comment]
@kwea123 
The reward is given when the player touches the next box. If steps are taken when the player is in the air, the action and reward won't be corresponding.

**Comment by ggbaird** [comment]
Hi @awjuliani,

Has there been any progress on selective decision/action making?

**Comment by awjuliani** [comment]
Hi @ggbaird,

We have actually implemented this, and are calling it "On Demand Decision Making." Please use the `development-0.3` branch if you'd like to try it out. https://github.com/Unity-Technologies/ml-agents/blob/development-0.3/docs/Learning-Environment-On-Demand-Decision.md

**Comment by unityjeffrey** [comment]
Hi all - closing this issue out.  On-demand decision making has been incorporated into Master and released as part of 0.3.  However, feel free to reopen the issue (or create a new issue) if this does not resolve your issue.
